caml_z.c

```diff
@@ -637,22 +639,23 @@ CAMLprim value ml_z_of_substring_base(value b, value v, value offset, value leng
 #endif
   {
      /* converts to sequence of digits */
-    char* dd = (char*)malloc(sz+1);
-    strncpy(dd,d,sz);
-    /* make sure that dd is nul terminated */
-    dd[sz] = 0;
-    for (i = 0; i < sz; i++) {
-      if (dd[i] >= '0' && dd[i] <= '9') dd[i] -= '0';
-      else if (dd[i] >= 'a' && dd[i] <= 'f') dd[i] -= 'a' - 10;
-      else if (dd[i] >= 'A' && dd[i] <= 'F') dd[i] -= 'A' - 10;
+    char* digits = (char*)malloc(sz+1);
+    for (i = 0, j = 0; i < sz; i++, j++) {
+      /* skip underscores but leading ones */
+      if (i > 0 && d[i] == '_') {j--;continue;};
```

Comment:
It's a matter of taste, but I would rather not increment `j` in the "next" part of the `for`, rather at the end of the body:
```
    for (i = 0, j = 0; i < sz; i++) {
      /* skip underscores but leading ones */
      if (i > 0 && d[i] == '_') continue;
      ....
      j++;
    }
```